panacea/query_client.go

```diff
@@ -98,46 +103,108 @@ func NewQueryClient(ctx context.Context, config *config.Config, info TrustedBloc
 		pvs,
 		store,
 		light.SkippingVerification(light.DefaultTrustLevel),
-		light.Logger(log.TestingLogger()),
+		light.Logger(tmlog.TestingLogger()),
 	)
 	if err != nil {
 		return nil, err
 	}
 
+	// call refresh every minute
+	go func() {
+		for {
+			time.Sleep(1 * time.Minute)
+			if err := refresh(ctx, lc, trustOptions.Period, &lcMutex); err != nil {
+				log.Errorf("light client refresh error: %v", err)
+			}
+		}
+	}()
+
 	return &QueryClient{
 		RpcClient:         rpcClient,
 		LightClient:       lc,
 		interfaceRegistry: makeInterfaceRegistry(),
 		sgxLevelDB:        db,
+		mutex:             &lcMutex,
 	}, nil
 }
 
+// refresh update light block, if the last light block has been updated more than trustPeriod * 2/3 ago.
+func refresh(ctx context.Context, lc *light.Client, trustPeriod time.Duration, m *sync.Mutex) error {
+	log.Info("check latest light block")
+	lastBlockHeight, err := lc.LastTrustedHeight()
+	if err != nil {
+		return err
+	}
+	lastBlock, err := lc.TrustedLightBlock(lastBlockHeight)
+	if err != nil {
+		return err
+	}
+	lastBlockTime := lastBlock.Time
+	currentTime := time.Now()
+	timeDiff := currentTime.Sub(lastBlockTime)
+	if timeDiff > trustPeriod*2/3 {
+		log.Info("update latest light block")
+		m.Lock()
+		if _, err := lc.Update(ctx, time.Now()); err != nil {
+			return err
+		}
+		m.Unlock()
+	}
+	return nil
+}
+
 // GetStoreData get data from panacea with storeKey and key, then verify queried data with light client and merkle proof.
 // the returned data type is ResponseQuery.value ([]byte), so recommend to convert to expected type
 func (q QueryClient) GetStoreData(ctx context.Context, storeKey string, key []byte) ([]byte, error) {
+	var queryHeight int64
+
+	// get recent light block
+	// if the latest block has already been updated, get LastTrustedHeight
+	q.mutex.Lock()
 	trustedBlock, err := q.LightClient.Update(ctx, time.Now())
+	q.mutex.Unlock()
```

Comment:
The best practice of unlocking the mutex is using `defer q.mutex.Unlock()`, so that the mutex can be unlocked in any cases regardless what happened in this function.

You may think that it's safe to unlock it right after calling the `q.LightClient.Update(...)` (aka. [critical section](https://en.wikipedia.org/wiki/Critical_section)). But, the critical section can raise a [panic](https://go.dev/blog/defer-panic-and-recover). Then, this `GetStoreData()` function will be returned, and the panic will be thrown to the caller of `GetStoreData()`. If so, the `q.mutex.Unlock()` won't be executed because the process won't reach to the line 165. 
In our current codes, your code could be safe because we aren't recovering any panic in our implementation. But, we can probably do it in the future. Who knows. 

So, it's much better to use `defer` to unlock mutexes safely. The `defer` statement is executed even if a panic occurs. You can see my example: https://gist.github.com/youngjoon-lee/8c3304b556ac6ce6a3ef856a91950f16.

Because the `GetStoreData()` function is big, I would recommend you to define another function which just calls `q.LightClient.Update()` with mutex.
```go
func (q QueryClient) safeUpdateLightClient(...) {
  q.mutex.Lock()
  defer q.mutex.Unlock()

  return q.LightClient.Update(...)
}
```

You can use this practice for the `q.LightClient.VerifyLightBlockAtHeight` at line 195 as well.

Reply:
Thanks! I fixed it in [0a8904b](https://github.com/medibloc/panacea-doracle/pull/21/commits/0a8904ba57ea69fd3fd4694cc5779fa85f47d8f3). I did not considered panic, thanks for let me know.